Fix Async schema loading so schemas cannot be used until they are completely loaded (#5441)
iTwin/itwinjs-core: docs/changehistory/NextVersion.md

}
```

### Interfaces changed

#### @itwin/core-quantity

- The interface `UnitConversion` has been renamed to [UnitConversionProps]($quantity).

#### @itwin/ecschema-metadata

- Renamed Interfaces
  - The interface `FormatProps` has been changed to a type alias [SchemaItemFormatProps]($ecschema-metadata).
  - The interface `UnitProps` has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
- `ISchemaLocater`
  - `getSchema` and `getSchemaSync` methods now take a `Readonly<SchemaKey>` instead of a `SchemaKey` and the `SchemaContext` parameter is required

>   - Added `getSchemaInfo` method to facilitate asynchronous schema loading

Reviewer: - The `FormatProps` interface has been replaced with the [SchemaItemFormatProps]($ecschema-metadata) type alias.
- The `UnitProps` interface has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
- [ISchemaLocater.getSchema]($ecschema-metadata) and [ISchemaLocater.getSchemaSync]($ecschema-metadata) now take a `Readonly<SchemaKey>` instead of a [SchemaKey]($ecschema-metadata) and the [SchemaContext]($ecschema-metadata) parameter is no longer optional.
- Added [ISchemaLocater.getSchemaInfo]($ecschema-metadata) to facilitate asynchronous schema loading.

> - Renamed Interfaces
>   - The interface `FormatProps` has been changed to a type alias [SchemaItemFormatProps]($ecschema-metadata).
>   - The interface `UnitProps` has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
> - `ISchemaLocater`
>   - `getSchema` and `getSchemaSync` methods now take a `Readonly<SchemaKey>` instead of a `SchemaKey` and the `SchemaContext` parameter is required
>   - Added `getSchemaInfo` method to facilitate asynchronous schema loading

Reviewer: ```suggestion
- The `FormatProps` interface has been replaced with the [SchemaItemFormatProps]($ecschema-metadata) type alias.
- The `UnitProps` interface has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
- [ISchemaLocater.getSchema]($ecschema-metadata) and [ISchemaLocater.getSchemaSync]($ecschema-metadata) now take a `Readonly<SchemaKey>` instead of a [SchemaKey]($ecschema-metadata) and the [SchemaContext]($ecschema-metadata) parameter is no longer optional.
```

```ts
const context = new SchemaContext();
const locater = new SchemaXmlFileLocater();
locater.addSchemaSearchPath("/Users/me/schemas/");
context.addLocater(locater);

const schemaKey = new SchemaKey("MySchemaWithManyReferences", 1, 0, 42);

// Start loading the schema but return as soon as we have loaded the name and version of the schema and it's references
const schemaInfo = await context.getSchemaInfo(schemaKey, SchemaMatchType.Exact);
// Get the whole schema either awaiting the schema promise created by getSchemaInfo or start loading if not already started
const schema = await context.getSchema(schemaKey, SchemaMatchType.Exact);
// Await the schema promise created by getSchemaInfo or return undefined if not already started
const schema2 = await context.getCachedSchema(schemaKey, SchemaMatchType.Exact);
```

### Other minor API changes

- Added `SchemaInfo` interface with schema keys for a schema and it's references.  `Schema` implicitly supports this interface.
- Some beta components had breaking changes and were moved to internal:
  - `SchemaGraph`
    - Now supports working with a `SchemaInfo` and a `SchemaContext` necessitating the init be made async.
  - `SchemaMap`

>     - Use `Array<Schema>` in it's place.

Reviewer: ```suggestion
    - Use `Array<Schema>` in its place.
```

  - `SchemaCache`
    - Updated to support caching partially loaded schemas, use `SchemaContext` to cache schemas in it's place.

> - Added helper method `SchemaFileUtility.writeSchemaToXmlString` to write schema xml to a string ``

Reviewer 2: ```suggestion
- Added helper method `SchemaFileUtility.writeSchemaToXmlString` to write schema xml to a string
```
